src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,846 @@
+        }
+
+        protected override Batch CreateBatch(DataViewRowCursor input) => new Batch(_batchSize, _outputLength, _modeler);
+
+        protected override Func<bool> GetIsNewBatchDelegate(DataViewRowCursor input)
+        {
+            return () => _batchSize == -1 ? input.Position == 0 : input.Position % _batchSize == 0;
+        }
+
+        protected override Func<bool> GetLastInBatchDelegate(DataViewRowCursor input)
+        {
+            return () => _batchSize == -1 ? input.Position == -1 : (input.Position + 1) % _batchSize == 0;
+        }
+
+        protected override ValueGetter<double> GetLookAheadGetter(DataViewRowCursor input)
+        {
+            return input.GetGetter<double>(input.Schema[_inputColumnName]);
+        }
+
+        protected override Func<int, bool> GetSchemaBindingDependencies(Func<int, bool> predicate)
+        {
+            return _bindings.GetDependencies(predicate);
+        }
+
+        protected override void ProcessExample(Batch currentBatch, double currentInput)
+        {
+            currentBatch.AddValue(currentInput);
+        }
+
+        protected override void ProcessBatch(Batch currentBatch)
+        {
+            currentBatch.Process();
+            currentBatch.Reset();
+        }
+
+        internal sealed class Batch
+        {
+            private List<double> _previousBatch;
+            private List<double> _batch;
+            private readonly int _outputLength;
+            private SrCnnEntireModeler _modeler;
+            private int _batchSize;
+            private double[][] _results;
+
+            public Batch(int batchSize, int outputLength, SrCnnEntireModeler modeler)
+            {
+                _batchSize = batchSize;
+                _outputLength = outputLength;
+                if (batchSize == -1)
+                {
+                    _previousBatch = new List<double>();
+                    _batch = new List<double>();
+                }
+                else
+                {
+                    _previousBatch = new List<double>(batchSize);
+                    _batch = new List<double>(batchSize);
+                }
+                _modeler = modeler;
+            }
+
+            public void AddValue(double value)
+            {
+                _batch.Add(value);
+            }
+
+            public int Count => _batch.Count;
+
+            public void Process()
+            {
+                _batchSize = _batch.Count;
+                if (_batch.Count < MinBatchSize)
+                {
+                    if (_previousBatch.Count + _batch.Count < MinBatchSize)
+                        return;
+                    var bLen = _previousBatch.Count - _batch.Count;
+                    _previousBatch = _previousBatch.GetRange(_batch.Count, bLen);
+                    _previousBatch.AddRange(_batch);
+                    _modeler.Train(_previousBatch.ToArray(), ref _results);
+                    _results = _results.Skip(bLen).ToArray();
```

Comment:
>Skip(bLen).ToArray() [](start = 40, length = 20)

Suggestion: Instead of skipping, keep an `int` indicating the "beginning" of `_results`, so the getter would return the value in `_results[input.Position % _batchSize + _bLen]` instead of `_results[input.Position % _batchSize]`.

Reply:
Fixed as you suggested.

---
In reply to: [427863039](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427863039) [](ancestors = 427863039)